shared/actions/platform-specific/index.native.tsx

```diff
@@ -458,6 +462,86 @@ function* requestContactPermissions(
   ])
 }
 
+async function manageContactsCache(
+  state: TypedState,
+  action: SettingsGen.LoadedContactImportEnabledPayload | ConfigGen.MobileAppStatePayload,
+  logger: Saga.SagaLogger
+) {
+  if (action.type === ConfigGen.mobileAppState && action.payload.nextAppState !== 'active') {
+    return
+  }
+
+  if (state.settings.contacts.importEnabled === false) {
+    return RPCTypes.contactsSaveContactListRpcPromise({contacts: []}).then(() =>
+      SettingsGen.createSetContactImportedCount({count: null})
+    )
+  }
+
+  // get permissions if we haven't loaded them for some reason
+  let {permissionStatus} = state.settings.contacts
+  if (permissionStatus === 'unknown') {
+    permissionStatus = await loadContactPermissionFromNative()
+  }
+  const perm = permissionStatus === 'granted'
+
+  const enabled = state.settings.contacts.importEnabled
+  if (!enabled || !perm) {
+    if (enabled && !perm) {
+      logger.info('contact import enabled but no contact permissions')
+    }
+    if (enabled === null) {
+      logger.info("haven't loaded contact import enabled")
+    }
+    return
+  }
+
+  // feature enabled and permission granted
+  const contacts = await Contacts.getContactsAsync()
+  const mapped = contacts.data.reduce((ret: Array<RPCTypes.Contact>, contact) => {
+    const {name, phoneNumbers = [], emails = []} = contact
+
+    const components = phoneNumbers.reduce(
+      (res, pn) => {
+        const formatted = getE164(pn.countryCode, pn.number)
+        if (formatted) {
+          res.push({
+            label: pn.label,
+            phoneNumber: phoneUtil.parse(pn.countryCode + pn.number),
+          })
+        }
+        return res
+      },
+      [] as Array<RPCTypes.ContactComponent>
+    )
+
+    components.push(...emails.map(e => ({label: e.label, email: e.email})))
+    if (components.length) {
+      ret.push({name, components})
+    }
+
+    return ret
+  }, [])
+  logger.info(`Importing ${mapped.length} contacts.`)
+  return RPCTypes.contactsSaveContactListRpcPromise({contacts: mapped})
+    .then(() => {
+      logger.info(`Success`)
+      return SettingsGen.createSetContactImportedCount({count: mapped.length})
+    })
+    .catch(e => {
+      logger.error('Error saving contacts list: ', e.message)
+    })
+}
+
+// Get phone number in e.164, or null if we can't parse it.
+const getE164 = (countryCode: string, phoneNumber: string) => {
+  try {
+    const parsed = phoneUtil.parse(countryCode + phoneNumber)
+    return phoneUtil.format(parsed, PhoneNumberFormat.E164) as string
```

Comment:
why need to cast this?

Reply:
`phoneUtil` is typed as `any`, from libphonenumber

Reply:
i thought there was a pr to pull this in: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/google-libphonenumber/index.d.ts

Reply:
yep by Jakob who's ooo this week. I'll take over his PR and update this once the types are in.